httpbakery/form/client.go

```diff
@@ -0,0 +1,201 @@
+package form
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/juju/schema"
+	"golang.org/x/crypto/ssh/terminal"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/environschema.v1"
+)
+
+// PromptingFiller implements a Filler by prompting for each field in
+// turn.
+type PromptingFiller struct {
+	// The Prompter to use to get the responses if this is nil then
+	// DefaultPrompter will be used.
+	Prompter Prompter
+}
+
+// Fill processes fields by first sorting them and then prompting for the
+// value of each one in turn.
+//
+// The fields are sorted by first by group name. Those in the same group
+// are sorted so that secret fields come after non-secret ones, finally
+// the fields are sorted by description.
+//
+// Each field will be prompted for, the returned value will then be
+// validated against the type for the field. If the returned value does
+// not validate correctly it will be prompted again twice before giving
+// up.
+//
+// If a field contains environment variables these will be checked in
+// order. The first one that is set will be used as the value of default
+// when Prompt is called.
+func (f PromptingFiller) Fill(fields environschema.Fields) (map[string]interface{}, error) {
+	fs := make(fieldSlice, len(fields))
```

Comment:
My preferred idiom for this kind of thing is:

    fs := make(fieldSlice, 0, len(fields))

then use append in the loop. Then you don't have to maintain the index.